### PR TITLE
[Feature] Custom wordings

### PR DESCRIPTION
**Pull request by jorGonzalez291292**
## Motivación y Contexto
It is necessary to have more contextual texts about what things can or cannot be done with each payment method.

## Descripción
Models that help us to customize the wordings for the modals of the behaviors are added to the PX signature as a payment method selector. These are applied through rules defined by the integrator.

[Documentation](https://docs.google.com/document/d/15Rlj5hJqfsZR49_wsipT41XRYUmfmRHQBmn70EazjgU/edit?usp=sharing)

### CODE REVIEW

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/di/MapperProvider.kt [hunk 1]
```diff
@@ -20,12 +21,13 @@ import com.mercadopago.android.px.internal.features.payment_result.remedies.Alte
 import com.mercadopago.android.px.internal.features.payment_result.remedies.RemediesLinkableMapper
 import com.mercadopago.android.px.internal.features.security_code.RenderModeMapper
 import com.mercadopago.android.px.internal.features.security_code.mapper.BusinessSecurityCodeDisplayDataMapper
+import com.mercadopago.android.px.internal.mappers.*
```
Comment (RoBarriosG) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/di/MapperProvider.kt:24]: minor comment, remove wildcard imports

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/one_tap/slider/PaymentMethodFragment.java [hunk 2]
```diff
@@ -304,7 +311,17 @@ public void enable() {
...
     public void onAction(@NonNull final GenericDialogAction genericDialogAction) {
-        //Do nothing
+        if(genericDialogAction instanceof GenericDialogAction.DeepLinkAction) {
+            startDeepLink(((GenericDialogAction.DeepLinkAction) genericDialogAction).getDeepLink());
+        }
+    }
+
+    private void startDeepLink(@NonNull final String deepLink) {
+        try {
+            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(deepLink)));
```
Comment (RoBarriosG) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/features/one_tap/slider/PaymentMethodFragment.java:321]: Use MercadoPagoUtil.getIntent(deepLink)

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/mappers/InitRequestBodyMapper.kt [hunk 3]
```diff
@@ -8,15 +8,13 @@ import com.mercadopago.android.px.internal.extensions.isZero
 import com.mercadopago.android.px.internal.features.FeatureProvider
 import com.mercadopago.android.px.internal.repository.PaymentSettingRepository
 import com.mercadopago.android.px.internal.tracking.TrackingRepository
-import com.mercadopago.android.px.model.internal.CheckoutFeaturesDM
-import com.mercadopago.android.px.model.internal.DiscountParamsConfigurationDM
-import com.mercadopago.android.px.model.internal.InitRequestBody
-import com.mercadopago.android.px.model.internal.PaymentTypeChargeRuleDM
+import com.mercadopago.android.px.model.internal.*
```
Comment (RoBarriosG) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/mappers/InitRequestBodyMapper.kt:11]: remove wildcard imports

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/one_tap/slider/PaymentMethodFragment.java [hunk 4]
```diff
@@ -304,7 +311,17 @@ public void enable() {
 
     @Override
     public void onAction(@NonNull final GenericDialogAction genericDialogAction) {
-        //Do nothing
+        if(genericDialogAction instanceof GenericDialogAction.DeepLinkAction) {
```
Comment (francofusaro) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/features/one_tap/slider/PaymentMethodFragment.java:314]: missing space between if and "("

File: px-services/build.gradle [hunk 5]
```diff
@@ -27,7 +27,7 @@ android {
     buildTypes {
         debug {
             buildConfigField "String", "API_ENVIRONMENT", "$Environment.LEGACY_PROD"
-            buildConfigField "String", "API_ENVIRONMENT_NEW", "$Environment.PRODUCTION"
+            buildConfigField "String", "API_ENVIRONMENT_NEW", "$Environment.GAMMA"
```
Comment (francofusaro) [inline px-services/build.gradle:30]: Watch out! 😛

File: px-services/src/main/java/com/mercadopago/android/px/model/internal/InitRequestBody.kt [hunk 6]
```diff
@@ -10,6 +10,8 @@ data class InitRequestBody(
     @SerializedName("discount_configuration")
     val discountParamsConfiguration: DiscountParamsConfigurationDM,
     val features: CheckoutFeaturesDM,
+    @SerializedName("payment_method_behaviour")
```
Comment (francofusaro) [inline px-services/src/main/java/com/mercadopago/android/px/model/internal/InitRequestBody.kt:13]: Can this be in plural? 😄

File: px-services/src/main/java/com/mercadopago/android/px/model/internal/Application.kt [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
+import com.google.gson.annotations.SerializedName
 import com.mercadopago.android.px.model.StatusMetadata
+import com.mercadopago.android.px.model.one_tap.CheckoutBehaviour
 
 data class Application(
     val paymentMethod: PaymentMethod,
     val validationPrograms: List<ValidationProgram>?,
-    val status: StatusMetadata
+    val status: StatusMetadata,
+    @SerializedName("behaviour")
```
Comment (francofusaro) [inline px-services/src/main/java/com/mercadopago/android/px/model/internal/Application.kt:11]: Can this be in plural? 😄

File: px-services/src/main/java/com/mercadopago/android/px/model/internal/PaymentMethodBehaviourDM.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.mercadopago.android.px.model.internal
+
+import com.google.gson.annotations.SerializedName
+
+data class PaymentMethodBehaviourDM(
+    @SerializedName("payment_types_rules")
```
Comment (francofusaro) [inline px-services/src/main/java/com/mercadopago/android/px/model/internal/PaymentMethodBehaviourDM.kt:6]: Can "types" be in singular? 😄

**Review by francofusaro — CHANGES_REQUESTED** [review]

**Comment by jorGonzalez291292** [comment]
@francofusaro They have to change the backend, for the moment to test I use the serializationName(""). When the changes are updated I'll update it.

**Review by francofusaro — APPROVED** [review]

**Review by daniivillegas — APPROVED** [review]

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,6 @@
 ## Unreleased
 _XX_04_2022_
+* FEATURE - Custom behaviour
```
Comment (guchito9) [inline CHANGELOG.md:3]: It lacks some context, custom behaviour for what?

File: px-checkout/src/main/java/com/mercadopago/android/px/configuration/AdvancedConfiguration.java [hunk 10]
```diff
@@ -333,6 +341,17 @@ public Builder setPaymentMethodRuleSet(@NonNull @PaymentMethodRules final List<
             return this;
         }
 
+        /**
+         * Provides additional settings to modify the types of payment methods to display at checkout.
+         *
+         * @param paymentMethodBehaviours your configuration.
+         * @return builder to keep operating
+         */
+        public Builder setPaymentMethodBehaviours(final List<PaymentMethodBehaviour> paymentMethodBehaviours) {
```
Comment (guchito9) [inline px-checkout/src/main/java/com/mercadopago/android/px/configuration/AdvancedConfiguration.java:350]: @NonNull?

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/disable_payment_method/DisabledPaymentMethodDetailDialog.java [hunk 11]
```diff
@@ -9,13 +9,15 @@
...
 import androidx.fragment.app.FragmentActivity;
+import com.mercadolibre.android.ui.widgets.MeliButton;
 import com.mercadolibre.android.ui.widgets.MeliDialog;
 import com.mercadopago.android.px.R;
 import com.mercadopago.android.px.internal.di.Session;
 import com.mercadopago.android.px.internal.util.TextUtil;
 import com.mercadopago.android.px.model.Payment;
 import com.mercadopago.android.px.model.StatusMetadata;
 import com.mercadopago.android.px.tracking.internal.views.DisabledPaymentMethodDetailViewTracker;
+import java.security.MessageDigest;
```
Comment (guchito9) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/features/disable_payment_method/DisabledPaymentMethodDetailDialog.java:20]: 👀

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/generic_modal/FromModalToGenericDialogItem.kt [hunk 12]
```diff
@@ -4,16 +4,21 @@ import com.mercadopago.android.px.internal.viewmodel.TextLocalized
...
+            value.dialogDescription,
+            modal.imageUrl,
+            TextLocalized(modal.title, 0),
+            TextLocalized(modal.description, 0),
+            modal.mainButton.let { Actionable(it.label, it.target, value.action) },
+            modal.secondaryButton?.let { Actionable(it.label, it.target, value.action) })
     }
 }
+
+data class FromModalToGenericDialogItemParams(
```
Comment (guchito9) [inline px-checkout/src/main/java/com/mercadopago/android/px/internal/features/generic_modal/FromModalToGenericDialogItem.kt:20]: I think it's better to simply name the class as Params but declare it as inner class of the mapper

File: px-services/src/main/java/com/mercadopago/android/px/model/internal/Application.kt [hunk 13]
```diff
@@ -1,11 +1,15 @@
...
+    val behaviours: Map<String, CheckoutBehaviour>
```
Comment (guchito9) [inline px-services/src/main/java/com/mercadopago/android/px/model/internal/Application.kt:12]: Should this be nullable or will come always from backend?

Comment (jorGonzalez291292) [reply px-services/src/main/java/com/mercadopago/android/px/model/internal/Application.kt:12]: This arrives empty from backend.

File: gradle/deploy_base.gradle [hunk 14]
```diff
@@ -84,8 +84,8 @@ afterEvaluate {
                 name = 'experimental'
                 url = 'https://android.artifacts.furycloud.io/repository/experimental'
                 credentials {
-                    username = "$System.env.NEXUS_DEPLOYER_USER"
-                    password = "$System.env.NEXUS_DEPLOYER_PASSWORD"
+                    username = "jorgonzalez"
+                    password = "n6J3E|_dt2_C18LoFmlv5rT||"
```
Comment (guchito9) [inline gradle/deploy_base.gradle:88]: 👀

**Review by guchito9 — APPROVED** [review]

**Pull request merged**